### PR TITLE
Improve `cannot move out of` error message

### LINKED ISSUE #99470: Error message mentions "x.0" (x is Option)
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=449ac104599c2bb219212dd16afa49d1

```rust
fn main() {
    let x: &Option<Box<i32>> = &Some(Box::new(0));

    match x {
        &Some(_y) => (),
        &None => (),
    }
}
```

The current output is:

```text
Compiling playground v0.0.1 (/playground)
error[[E0507]](https://doc.rust-lang.org/stable/error-index.html#E0507): cannot move out of `x.0` which is behind a shared reference
 --> src/main.rs:4:7
  |
4 | match x {
  |       ^
5 |     &Some(_y) => (),
  |     ---------
  |     |     |
  |     |     data moved here
  |     |     move occurs because `_y` has type `Box<i32>`, which does not implement the `Copy` trait
  |     help: consider removing the `&`: `Some(_y)`
```

The output mentions `x.0`, but `x` is `Option`, so `x.0` is nonsense

### PR DESCRIPTION
**Pull request by obeis**
Closes #99470
r? @bjorn3

### CODE REVIEW

File: src/test/ui/borrowck/borrow-tuple-fields.rs [hunk 1]
```diff
@@ -30,12 +30,12 @@ fn main() {
 
     let mut x = Bar(1, 2);
     let a = &x.0;
-    let b = &mut x.0; //~ ERROR cannot borrow `x.0` as mutable because it is also borrowed as
+    let b = &mut x.0; //~ ERROR cannot borrow `x` as mutable because it is also borrowed as
```
Comment (compiler-errors) [inline src/test/ui/borrowck/borrow-tuple-fields.rs:33]: what's wrong with `x.0` here? Isn't the point of #99470 that we should not say `x.0` if it's an enum destructuring, not getting rid of the `.0`/`.1`/etc. altogether?

**Comment by bjorn3** [comment]
Not familiar with the borrowck impl.

r? @compiler-errors Feel free to reassign.

File: src/test/ui/borrowck/borrow-tuple-fields.rs (hunk 1 above)
Comment (obeis) [reply src/test/ui/borrowck/borrow-tuple-fields.rs:33]: I am going to fix this

File: src/test/ui/moves/move-out-of-tuple-field.stderr [hunk 2]
```diff
@@ -8,15 +8,15 @@ LL |     let z = x.0;
    |
    = note: move occurs because `x.0` has type `Box<i32>`, which does not implement the `Copy` trait
 
-error[E0382]: use of moved value: `x.0`
+error[E0382]: use of moved value: `x`
```
Comment (compiler-errors) [inline src/test/ui/moves/move-out-of-tuple-field.stderr:11]: Hm, can you investigate why this is still a regression?

Comment (obeis) [reply src/test/ui/moves/move-out-of-tuple-field.stderr:11]: sure

File: compiler/rustc_borrowck/src/diagnostics/move_errors.rs [hunk 3]
```diff
@@ -368,13 +368,27 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
+                            including_downcast: false,
+                            including_tuple_field: false,
+                        },
+                    ),
+                    self.describe_name(move_place_ref),
+                    source.describe_for_named_place(),
+                ) {
+                    (Some(place_desc), Some(name), Some(source_desc)) => self.cannot_move_out_of(
+                        span,
+                        &format!("`{place_desc}` as `{name}` which is behind a {source_desc}"),
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/diagnostics/move_errors.rs:385]: maybe "as enum variant `{name}`"... unless this applies to something more than just enum variants?

**Review by compiler-errors — COMMENTED** [review]
@obeis, can you please squash all of your changes into one commit?

File: compiler/rustc_borrowck/src/diagnostics/mod.rs [hunk 4]
```diff
@@ -43,7 +43,12 @@ pub(crate) use region_name::{RegionName, RegionNameSource};
 pub(crate) use rustc_const_eval::util::CallKind;
 use rustc_middle::mir::tcx::PlaceTy;
 
-pub(super) struct IncludingDowncast(pub(super) bool);
+pub(super) struct DescribePlaceOpt {
+    pub including_downcast: bool,
+    pub including_tuple_field: bool,
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/diagnostics/mod.rs:48]: can you describe what `including_tuple_field` does?

File: compiler/rustc_borrowck/src/diagnostics/move_errors.rs [hunk 5]
```diff
@@ -368,13 +368,27 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
+                        },
+                    ),
+                    self.describe_name(move_place_ref),
+                    source.describe_for_named_place(),
+                ) {
+                    (Some(place_desc), Some(name), Some(source_desc)) => self.cannot_move_out_of(
+                        span,
+                        &format!("`{place_desc}` as enum variant `{name}` which is behind a {source_desc}"),
+                    ),
+                    (Some(place_desc), _, Some(source_desc)) => self.cannot_move_out_of(
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/diagnostics/move_errors.rs:387]: Do you want to handle the `Some(place_desc), Some(name), None` case?

**Comment by compiler-errors** [comment]
One last comment and then I think this might ready to land.

File: compiler/rustc_borrowck/src/diagnostics/move_errors.rs (hunk 5 above)
Comment (obeis) [reply compiler/rustc_borrowck/src/diagnostics/move_errors.rs:387]: Yes

**Comment by compiler-errors** [comment]
@obeis please rebase your changes instead of using a merge commit

**Comment by obeis** [comment]
@compiler-errors fixed!

**Pull request merged**